File: typescript/src/sign.ts
import { autograph_sign_subject } from './clib'
import { SignFunction } from '../../types'
import { createSignatureBytes } from './utils'
import { SigningError } from './error'

const createSign =
  (identityPrivateKey: Uint8Array): SignFunction =>
  (subject: Uint8Array) => {
    const signature = createSignatureBytes()
    const success =
      autograph_sign_subject(
        signature,
        identityPrivateKey,
        subject,
        subject.byteLength
      ) === 0
    if (!success) {
      throw new SigningError()
    }
    return signature
  }

export default createSign

File: rust/src/sign.rs
use alloc::boxed::Box;
use alloc::vec::Vec;

use crate::clib::autograph_sign_subject;
use crate::error::Error;
use crate::utils::create_signature_bytes;

pub type SignFunction = Box<dyn Fn(&Vec<u8>) -> Result<Vec<u8>, Error>>;

pub fn create_sign(identity_private_key: Vec<u8>) -> SignFunction {
    Box::new(move |subject: &Vec<u8>| {
        let mut signature = create_signature_bytes();
        let success = unsafe {
            autograph_sign_subject(
                signature.as_mut_ptr(),
                identity_private_key.as_ptr(),
                subject.as_ptr(),
                subject.len() as u32,
            )
        } == 0;
        if !success {
            Err(Error::Signing)
        } else {
            Ok(signature)
        }
    })
}
